### PR TITLE
GPU pixel local reconstruction: make the pixel cluster thresholds configurable

### PR DESCRIPTION
**Pull request by czangela**
#### PR description:

This is an effort towards solving [one of the issues](https://github.com/cms-sw/cmssw/issues/32483#issuecomment-745388762) in #32483.

Hardcoded charge cuts on pixel clusters were used in:  
`RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterChargeCut.h` and  
`RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc`

Since the GPU version uses these thresholds in the free function `gpuClusterChargeCut` nested in the `gpuClustering` namespace, they are propagated through function parameters, passing them by value. They are initially set in the `SiPixelRawToClusterCUDA` class.

The `SiPixelDigisClustersFromSoA` has these parameters as private members (as does the legacy version).

#### PR validation:

These changes should not affect validation results.  
Performance should not be affected.

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

### CODE REVIEW

**Comment by slava77** [comment]
enable gpu

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterThresholds_h
+#define RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterThresholds_h
+
+struct gpuClusterThresholds {
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:4]: "G"

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterThresholds_h
```
Comment (davidlange6) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:1]: looks like this needs to be in the interface directory

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+#define RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterThresholds_h
+
+struct gpuClusterThresholds {
+  constexpr gpuClusterThresholds(int32_t clusterThreshold_L1, int32_t clusterThreshold)
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:5]: not needd

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc [hunk 4]
```diff
@@ -89,7 +91,9 @@ SiPixelRawToClusterCUDA::SiPixelRawToClusterCUDA(const edm::ParameterSet& iConfi
           edm::ESInputTag("", iConfig.getParameter<std::string>("CablingMapLabel")))),
       isRun2_(iConfig.getParameter<bool>("isRun2")),
       includeErrors_(iConfig.getParameter<bool>("IncludeErrors")),
-      useQuality_(iConfig.getParameter<bool>("UseQualityInfo")) {
+      useQuality_(iConfig.getParameter<bool>("UseQualityInfo")),
+      clusterThresholds_(iConfig.getParameter<int32_t>("theGPUClusterThreshold_L1"),
```
Comment (VinInn) [inline RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc:95]: use brace.

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h (hunk 2 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:1]: why?

**Comment by davidlange6** [comment]
> On Mar 24, 2021, at 5:01 PM, Vincenzo Innocente ***@***.***> wrote:
> 
> 
> @VinInn commented on this pull request.
> 
> In RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:
> 
> > @@ -0,0 +1,11 @@
> +#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_gpuClusterThresholds_h
> 
> why?

Its used somewhere other than this directory

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h (hunk 2 above)
Comment (VinInn) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h:1]: test used not to count.

**Comment by makortel** [comment]
> > why?
>
> Its used somewhere other than this directory

In practice we allow header-only code from `plugins` to be included in `test` of the same package.

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc (hunk 4 above)
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc]: done

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h (hunk 3 above)
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h]: done

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h (hunk 1 above)
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h]: done

**Comment by davidlange6** [comment]
Ok - I didn’t realize. Thx

File: RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h (hunk 2 above)
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/gpuClusterThresholds.h]: This wasn't quite clear to me, so it's good that it got cleared up. :)

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
@cmsbuild please test

now with GPU tests and comparisons in place

**Comment by fwyzard** [comment]
While I don't expect differences from these changes, can you compare the performance (event throughput and/or reconstruction time) before and after these changes ?

**Comment by fwyzard** [comment]
By the way, can something automatic be set up for the performance comparison ?

**Comment by slava77** [comment]
> While I don't expect differences from these changes, can you compare the performance (event throughput and/or reconstruction time) before and after these changes ?

is this a question to @czangela ?

**Comment by fwyzard** [comment]
> > While I don't expect differences from these changes, can you compare the performance (event throughput and/or reconstruction time) before and after these changes ?
> 
> is this a question to @czangela ?

Yep.

**Comment by slava77** [comment]
> By the way, can something automatic be set up for the performance comparison ?

what would be a good metric, especially in the constraints of the number of events that run in the full MC workflow (10)?
Perhaps some of the tools used in cms-patatrack can be added to the GPU tests. My guess is that for something sensible we'll need a workflow with substantially more events and perhaps without the validation/MC overhead.

From what's available in the jenkins jobs we have FW job report. Here step3 has
`<Metric Name="TotalLoopCPU" Value="7.68059"/> ` changing to 7.79762 for 10824.502.
and `TotalLoopTime 10.6184 -> 10.9811`; that's on 10 events. Assuming the total loop time also includes the GPU execution, there may be up to 0.25 s increase.
I doubt thought that the increment is related to the changes in this PR.

**Comment by fwyzard** [comment]
> My guess is that for something sensible we'll need a workflow with substantially more events and perhaps without the validation/MC overhead.

Agreed.
Once https://github.com/cms-sw/cmssw/pull/31722 is merged, a good starting point would running only the `step3` of `11634.502` with `--customise RecoTracker/Configuration/customizePixelOnlyForProfiling.customizePixelOnlyForProfilingGPUOnly` over 1k existing MC events (or something similar over Run 2 data).

>  I doubt thought that the increment is related to the changes in this PR.

Agreed ...

**Comment by fwyzard** [comment]
> Perhaps some of the tools used in cms-patatrack can be added to the GPU tests.

The tools are actually here: https://github.com/cms-patatrack/patatrack-scripts - and would probably need to be ade a bit more user friendly, but could be adapted for PR tests, of course.

However running a single job (maybe under [`ncu`](https://docs.nvidia.com/nsight-compute/NsightComputeCli/index.html)) might be enough.

**Review by slava77 — COMMENTED** [review]
some input on naming of the parameter structs from the DPG may be useful, especially if this ever makes it to a DB.

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc [hunk 5]
```diff
@@ -33,17 +33,25 @@ class SiPixelDigisClustersFromSoA : public edm::global::EDProducer<> {
 
   edm::EDPutTokenT<edm::DetSetVector<PixelDigi>> digiPutToken_;
   edm::EDPutTokenT<SiPixelClusterCollectionNew> clusterPutToken_;
+
+  //! Clustering-related quantities:
+  const int theClusterThreshold;     // Cluster threshold in electrons
+  const int theClusterThreshold_L1;  // Cluster threshold in electrons for Layer 1
```
[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelDigisClustersFromSoA.cc:39]: It's great that you bring this up, as I wanted to ask about this earlier. I thought that since the SoA version runs on the CPU, it should not use / know about data structures, constants defined for the GPU algorithms.
Parameters that are somehow connected are usually not grouped together (e.g. in helper structs) in CPU classes, so I thought that should stay like that.
However since the GPU version has to propagate these constants, it was a better choice, to use a helper struct.

File: RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h [hunk 6]
```diff
@@ -0,0 +1,9 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_GPUClusterThresholds_h
+#define RecoLocalTracker_SiPixelClusterizer_plugins_GPUClusterThresholds_h
+
+struct GPUClusterThresholds {
```
[the remark this answers is not in the document]
Comment (mmusich) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h:4]: Since we are at it we can remove `GPU` completely form the name.
Those constant have nothing to do with GPUs, and in fact are the same in the legacy code configuration:

https://github.com/cms-sw/cmssw/blob/9c0ef283f7847d7aca1e34791522637609dcee87/RecoLocalTracker/SiPixelClusterizer/python/SiPixelClusterizer_cfi.py#L15-L16

BTW, @slava77 if 

> some input on naming of the parameter structs from the DPG may be useful, especially if this ever makes it to a DB.

you were referring to this, I don't think we're planning to store this in DB. Once the new L1 is timed again and we know what threshold optimizes resolution we don't plan to change then again (TM)

File: RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+  const int32_t layer1;       // Cluster threshold in electrons for Layer 1
+  const int32_t otherLayers;  // Cluster threshold in electrons
+};
```
[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h]: Yes, but the whole point is that these are not constants, but rather they should be modifiable from config files.
Also, is this naming documented on http://cms-sw.github.io because I couldn't find it ?

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h [hunk 8]
```diff
@@ -13,6 +13,7 @@
 #include "HeterogeneousCore/CUDAUtilities/interface/host_noncached_unique_ptr.h"
 #include "DataFormats/SiPixelRawData/interface/SiPixelErrorCompact.h"
 #include "DataFormats/SiPixelRawData/interface/SiPixelFormatterErrors.h"
+#include "RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h"
```
[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h]: I've seen instances of both in the case of "local includes", I'm still not sure which one is preferable.

[the remark this answers is not in the document]
Comment (mmusich) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h]: I had the same curiosity. See explanation from Andrea here: https://github.com/cms-patatrack/cmssw/pull/593#discussion_r543555127

[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterGPUKernel.h]: I actually read that comment earlier and so was more afraid that people would start complaining about it if I did the same.

File: RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h]: done, and without the GPU suffix

File: RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (slava77) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h]: >  Also, is this naming documented on http://cms-sw.github.io because I couldn't find it ?

no, these are in the broad "common sense" category.

`k` is used as a prefix for (class/enum/global level constants) in different style guides, e.g. in google style https://google.github.io/styleguide/cppguide.html#Constant_Names

[the remark this answers is not in the document]
Comment (czangela) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h]: I know what it means, and also that it's in the Google C++ Style Guide. I also know of groups that follow the Google C++ Style Guide, but they documented it (I did not find the link, but they did the at DUNE). I mainly just meant, that if this is something I should use, I think it would useful to document it on the appropriate website.

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterThresholds.h [hunk 9]
```diff
@@ -0,0 +1,13 @@
+#ifndef RecoLocalTracker_SiPixelClusterizer_plugins_SiPixelClusterThresholds_h
+#define RecoLocalTracker_SiPixelClusterizer_plugins_SiPixelClusterThresholds_h
+
+struct SiPixelClusterThresholds {
+  SiPixelClusterThresholds() = default;
+  inline constexpr int32_t getThresholdForLayerOnCondition(bool isLayer1) const noexcept {
+    return isLayer1 ? kDefaultPhase1_layer1 : kDefaultPhase1_otherLayers;
+  }
+  const int32_t kDefaultPhase1_layer1 = 2000;
```
Comment (slava77) [inline RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelClusterThresholds.h:9]: this is rather confusing now; only `2000` is properly a default for phase1.
The actual configured values which are not `2000` are not a default anymore.

The reason I proposed the constants was to use the same value in SiPixelDigisClustersFromSoA.cc and SiPixelRawToClusterCUDA.cc fillDescriptions so that we have only one place to define constants which are supposed to have the same value.
However, the recent update introduced now the third repeated definition and made the naming somewhat more confusing.

The original names `layer1` and `otherLayers` was more appropriate for general use context.

File: RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc [hunk 10]
```diff
@@ -89,7 +91,9 @@ SiPixelRawToClusterCUDA::SiPixelRawToClusterCUDA(const edm::ParameterSet& iConfi
           edm::ESInputTag("", iConfig.getParameter<std::string>("CablingMapLabel")))),
       isRun2_(iConfig.getParameter<bool>("isRun2")),
       includeErrors_(iConfig.getParameter<bool>("IncludeErrors")),
-      useQuality_(iConfig.getParameter<bool>("UseQualityInfo")) {
+      useQuality_(iConfig.getParameter<bool>("UseQualityInfo")),
+      clusterThresholds_{iConfig.getParameter<int32_t>("theGPUClusterThreshold_L1"),
+                         iConfig.getParameter<int32_t>("theGPUClusterThreshold")} {
```
Comment (slava77) [inline RecoLocalTracker/SiPixelClusterizer/plugins/SiPixelRawToClusterCUDA.cc:96]:
```suggestion
      clusterThresholds_{iConfig.getParameter<int32_t>("clusterThreshold_layer1"),
                         iConfig.getParameter<int32_t>("clusterThreshold_otherLayers")} {
```
to match the naming in SiPixelDigisClustersFromSoA.cc (consistency is important, IMHO)

**Comment by slava77** [comment]
> Feel free to ignore this, but still - since we are suing camel case, would it make sense to use `kDefaultPhase1OtherLayers` instead of `kDefaultPhase1_otherLayers`, _etc_ ?

The reason I thought that a snake-camel case mix is exceptionally more appropriate to be able to more directly match that the (static) constants should correspond to default values of the non-const data members, while also being useful as constants outside of the `SiPixelClusterThresholds` definition.

Perhaps this conflict is an indication that the definition is in a wrong place and a more functional choice is to have a full object defined separately `constexpr kSiPixelClusterThresholdsDefaultPhase1 {.layer1 = 2000, .otherLayers = 4000};`

File: RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (slava77) [reply RecoLocalTracker/SiPixelClusterizer/plugins/GPUClusterThresholds.h]: we do not have anything close to a coherent style of constant names, perhaps it's worth to propose something (outside the scope of this PR discussion). 
In the context of this PR I chose to type something explicit instead of saying "please define a named constant".

File: RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2TrackerRecHits.cc [hunk 11]
```diff
@@ -1,31 +1,27 @@
...
-#include "DataFormats/TrackerRecHit2D/interface/Phase2TrackerRecHit1D.h"
-#include "DataFormats/TrackerCommon/interface/TrackerTopology.h"
-#include "DataFormats/DetId/interface/DetId.h"
-
-#include "RecoLocalTracker/Records/interface/TkPhase2OTCPERecord.h"
 #include "RecoLocalTracker/Phase2TrackerRecHits/interface/Phase2StripCPE.h"
-
-#include <vector>
-#include <string>
+#include "RecoLocalTracker/Records/interface/TkPhase2OTCPERecord.h"
```
Comment (mmusich) [inline RecoLocalTracker/Phase2TrackerRecHits/plugins/Phase2TrackerRecHits.cc:25]: why is this file changed?

File: RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc [hunk 12]
```diff
@@ -11,18 +11,19 @@
 // Magnetic field
 #include "MagneticField/Engine/interface/MagneticField.h"
 
-// Commented for now (3/10/17) until we figure out how to resuscitate 2D template splitter
+// Commented for now (3/10/17) until we figure out how to resuscitate 2D
+// template splitter
```
Comment (mmusich) [inline RecoLocalTracker/SiPixelRecHits/src/PixelCPEClusterRepair.cc:15]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/interface/PixelClusterSelectorTopBottom.h [hunk 13]
```diff
@@ -2,26 +2,23 @@
 #define RecoSelectors_PixelClusterSelectorTopBottom_h
 
 /* \class PixelClusterSelectorTopBottom
-*
-* \author Giuseppe Cerati, INFN
-*
-*
-*/
+ *
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/interface/PixelClusterSelectorTopBottom.h:5]: why is this file changed?

**Comment by fwyzard** [comment]
@czangela can you squash at least some of the commits, to avoid that a file is added and removed within the PR ?

File: RecoLocalTracker/SubCollectionProducers/src/HITrackClusterRemover.cc [hunk 14]
```diff
@@ -1,37 +1,32 @@
-#include "FWCore/Framework/interface/Frameworkfwd.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/HITrackClusterRemover.cc:1]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/HLTTrackClusterRemoverNew.cc [hunk 15]
```diff
@@ -1,35 +1,31 @@
-#include "FWCore/Framework/interface/Frameworkfwd.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/HLTTrackClusterRemoverNew.cc:1]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/JetCoreClusterSplitter.cc [hunk 16]
```diff
@@ -1,27 +1,23 @@
-#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include <algorithm>
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/JetCoreClusterSplitter.cc:1]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/PixelClusterSelectorTopBottom.cc [hunk 17]
```diff
@@ -1,6 +1,7 @@
-#include "FWCore/Framework/interface/MakerMacros.h"
 #include "RecoLocalTracker/SubCollectionProducers/interface/PixelClusterSelectorTopBottom.h"
 
+#include "FWCore/Framework/interface/MakerMacros.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/PixelClusterSelectorTopBottom.cc:3]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemover.cc [hunk 18]
```diff
@@ -1,32 +1,29 @@
-#include "FWCore/Framework/interface/Frameworkfwd.h"
-#include "FWCore/Framework/interface/stream/EDProducer.h"
-#include "FWCore/Framework/interface/Event.h"
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
-#include "FWCore/Utilities/interface/InputTag.h"
-
-#include "DataFormats/SiStripCluster/interface/SiStripCluster.h"
+#include "DataFormats/Common/interface/ContainerMask.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemover.cc:1]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc [hunk 19]
```diff
@@ -1,31 +1,26 @@
-#include "FWCore/Framework/interface/Frameworkfwd.h"
-#include "FWCore/Framework/interface/stream/EDProducer.h"
-#include "FWCore/Framework/interface/Event.h"
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
-#include "FWCore/Utilities/interface/InputTag.h"
-
+#include "DataFormats/Common/interface/ContainerMask.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/SeedClusterRemoverPhase2.cc:1]: why is this file changed?

File: RecoLocalTracker/SubCollectionProducers/src/StripClusterSelectorTopBottom.cc [hunk 20]
```diff
@@ -1,6 +1,7 @@
-#include "FWCore/Framework/interface/MakerMacros.h"
 #include "RecoLocalTracker/SubCollectionProducers/interface/StripClusterSelectorTopBottom.h"
 
+#include "FWCore/Framework/interface/MakerMacros.h"
```
Comment (mmusich) [inline RecoLocalTracker/SubCollectionProducers/src/StripClusterSelectorTopBottom.cc:3]: why is this file changed?

**Review by mmusich — COMMENTED** [review]
it seems there is a lot of unnecessary formatting going on.

**Comment by fwyzard** [comment]
please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/33259	822c264 
- code changes are in line with the PR description and the follow up review; the changes are not changing the default behavior.
- jenkins tests pass and comparisons with the baseline show no differences

**Comment by qliphy** [comment]
+1

**Comment by fwyzard** [comment]
### Comparison of the pixel-only reconstruction throughput, over Run2018D data at pileup 50

scan-136 885502
zoom-136 885502
